tez-runtime-library/src/main/java/org/apache/tez/runtime/library/common/sort/impl/IFile.java

```diff
@@ -71,6 +73,223 @@
+
+    private static final int checksumSize = IFileOutputStream.getCheckSumSize();
+
+    /**
+     * Note that we do not allow compression in in-mem stream.
+     * When spilled over to file, compression gets enabled.
+     *
+     * @param conf
+     * @param fs
+     * @param taskOutput
+     * @param keyClass
+     * @param valueClass
+     * @param codec
+     * @param writesCounter
+     * @param serializedBytesCounter
+     * @param cacheSize
+     * @throws IOException
+     */
+    public FileBackedInMemIFileWriter(Configuration conf, FileSystem fs,
+        TezTaskOutput taskOutput, Class keyClass, Class valueClass,
+        CompressionCodec codec,
+        TezCounter writesCounter,
+        TezCounter serializedBytesCounter,
+        int cacheSize) throws IOException {
+      super(conf, new FSDataOutputStream(createBoundedBuffer(cacheSize), null),
+          keyClass, valueClass, null, writesCounter, serializedBytesCounter);
+      this.fs = fs;
+      this.cacheStream = (BoundedByteArrayOutputStream) this.rawOut.getWrappedStream();
+      this.taskOutput = taskOutput;
+      this.bufferFull = (cacheStream == null);
+      this.cacheSize = (cacheStream == null) ? 0 :
+          Math.max(getBaseCacheSize(), cacheStream.available());
+      this.fileCodec = codec;
+    }
+
+    /**
+     * For basic cache size checks: header + checksum + EOF marker
+     *
+     * @return size of the base cache needed
+     */
+    static int getBaseCacheSize() {
+      return (checksumSize + (2 * WritableUtils.getVIntSize(EOF_MARKER))
+          + HEADER.length);
+    }
+
+    /**
+     * Create in mem stream. In it is too small, adjust it's size
+     *
+     * @param size
+     * @return in memory stream
+     */
+    public static BoundedByteArrayOutputStream createBoundedBuffer(int size) {
+      int resize = Math.max(getBaseCacheSize(), size);
+      return new BoundedByteArrayOutputStream(resize);
+    }
+
+    /**
+     * Flip over from memory to file based writer.
+     *
+     * 1. Content format: HEADER + real data + CHECKSUM. Checksum is for real
+     * data.
+     * 2. Before flipping, close checksum stream, so that checksum is written
+     * out.
+     * 3. Create relevant file based writer.
+     * 4. Write header and then real data.
+     *
+     * @throws IOException
+     */
+    private void resetToFileBasedWriter() throws IOException {
+      // Close out stream, so that data checksums are written.
+      // Buf contents = HEADER + real data + CHECKSUM
+      this.out.close();
+
+      // Get the buffer which contains data in memory
+      BoundedByteArrayOutputStream bout =
+          (BoundedByteArrayOutputStream) this.rawOut.getWrappedStream();
+
+      // Create new file based writer
+      if (outputPath == null) {
+        outputPath = taskOutput.getOutputFileForWrite();
+      }
+      LOG.info("Creating file " + outputPath);
+      FSDataOutputStream newRawOut = fs.create(outputPath);
```

Comment:
Most of this is identical to what is done in the IFile constructor. Think it makes sense to break this out into a separate method, and re-use it, instead of repeating the code.

Reply:
Yes, makes sense to create `setupOutputStream()` in `IFile` itself.